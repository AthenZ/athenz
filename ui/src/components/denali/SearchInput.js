/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import Input from './Input';
import Icon from './icons/Icon';
import { colors } from './styles/colors';

/**
 * This component is just a thin wrapper around
 * [`<Input>`](/#!/Input) that is styled for a search input field.
 */
// NOTE: This has to be a React class so refs work.
class SearchInput extends React.Component {
    render() {
        return (
            <Input
                {...this.props}
                renderIcon={({ sizePx }) => (
                    <Icon
                        icon={'search'}
                        color={colors.brand700}
                        size={sizePx}
                    />
                )}
            />
        );
    }
}

SearchInput.propTypes = {
    /** Additonal class to apply to the outer div */
    className: PropTypes.string,
    /** Dark theme */
    dark: PropTypes.bool,
    /** Force input into error state (similar to typing invalid input) */
    error: PropTypes.bool,
    /** Input fills entire width of parent container */
    fluid: PropTypes.bool,
    /** Attach a label to the right of the input field */
    label: PropTypes.any,
    /** Display a message under the input field */
    message: PropTypes.string,
    /** Placeholder (default: "Search") */
    placeholder: PropTypes.string,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Size (height) of input field */
    size: PropTypes.oneOf(['default', 'small']),
    /** Content of the textarea. This is the single source of truth */
    value: PropTypes.string,
};

SearchInput.defaultProps = {
    dark: false,
    disabled: false,
    error: false,
    fluid: false,
    noanim: false,
    placeholder: 'Search',
};

export default SearchInput;
