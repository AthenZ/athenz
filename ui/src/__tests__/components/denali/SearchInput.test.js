/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import SearchInput from '../../../components/denali/SearchInput';

describe('SearchInput', () => {
    it('should render', () => {
        const onChange = jest.fn();
        const { getByTestId } = render(
            <SearchInput
                name='search'
                placeholder='Search!'
                value=''
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('input-wrapper');
        const inputNode = getByTestId('input-node');

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.querySelector('.input-icon')).not.toBeNull();
        expect(inputNode).toHaveAttribute('placeholder', 'Search!');
    });

    it('should call onSearch when search icon is clicked', () => {
        const onChange = jest.fn();
        const onSearch = jest.fn();
        const { getByTestId } = render(
            <SearchInput
                name='search'
                placeholder='Search!'
                value=''
                onChange={onChange}
                onSearch={onSearch}
            />
        );
        const wrapper = getByTestId('input-wrapper');
        // Find the div that has the onClick handler (the one with cursor: pointer style)
        const searchIconContainer = wrapper.querySelector(
            '.input-icon[style*="cursor: pointer"]'
        );

        fireEvent.click(searchIconContainer);
        expect(onSearch).toHaveBeenCalledTimes(1);
    });

    it('should not call onSearch when search icon is clicked without onSearch prop', () => {
        const onChange = jest.fn();
        const { getByTestId } = render(
            <SearchInput
                name='search'
                placeholder='Search!'
                value=''
                onChange={onChange}
            />
        );
        const wrapper = getByTestId('input-wrapper');
        const searchIconContainer = wrapper.querySelector('.input-icon');

        // This should not throw an error
        expect(() => fireEvent.click(searchIconContainer)).not.toThrow();
    });
});
