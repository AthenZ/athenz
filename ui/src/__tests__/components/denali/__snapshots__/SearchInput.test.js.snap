// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput should render 1`] = `
.emotion-1 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  position: relative;
  width: 250px;
}

.emotion-1 input {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: rgba(53,112,244,0.05);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #303030;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: 36px;
  margin: 0;
  outline: none;
  padding: 0 1rem;
  text-align: left;
  width: 100%;
}

.emotion-1.animated input {
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out, border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out, border 0.2s ease-in-out;
}

.emotion-1 input::-webkit-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-1 input::-moz-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-1 input:-ms-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-1 input::placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-1 input.focused,
.emotion-1 input:active,
.emotion-1 input:focus {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #3570f4;
  color: #303030;
}

.emotion-1 input:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-1 .message {
  font-size: 12px;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

.emotion-1.error input,
.emotion-1.error input.focused,
.emotion-1.error input:active,
.emotion-1.error input:focus {
  border-bottom: 2px solid #d01111;
}

.emotion-1.error .message {
  color: #d01111;
}

.emotion-1 input {
  padding-right: 36px;
}

.emotion-1 .input-icon {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 36px;
  position: absolute;
  right: 6px;
}

.emotion-0 {
  fill: #3570f4;
  cursor: inherit;
  vertical-align: text-bottom;
}

<div
  class="emotion-1 denali-input animated"
  data-testid="input-wrapper"
>
  <input
    autocomplete="off"
    data-testid="input-node"
    name="search"
    placeholder="Search!"
    value=""
  />
  <div
    class="input-icon"
  >
    <svg
      class="emotion-0"
      data-testid="icon"
      height="24px"
      viewBox="0 0 1024 1024"
      width="24px"
    >
      <title>
        search
      </title>
      <path
        d="M659.413 599.68c40.125-53.054 64.27-120.136 64.27-192.857 0-177.556-143.937-321.493-321.493-321.493s-321.493 143.937-321.493 321.493c0 177.556 143.937 321.493 321.493 321.493 74.297 0 142.708-25.203 197.149-67.525l-0.726 0.543 267.307 264.96c7.723 7.731 18.396 12.514 30.187 12.514s22.464-4.782 30.186-12.513l0-0c7.795-7.733 12.621-18.45 12.621-30.293s-4.826-22.56-12.618-30.291l-0.003-0.003zM444.8 637.227c-11.829 2.113-25.446 3.322-39.345 3.322-129.603 0-234.667-105.064-234.667-234.667s105.064-234.667 234.667-234.667c129.603 0 234.667 105.064 234.667 234.667 0 13.899-1.208 27.516-3.526 40.752l0.204-1.407c-17.178 98.349-93.651 174.822-190.593 191.796l-1.407 0.204z"
      />
    </svg>
  </div>
  <div
    class="message"
    data-testid="message"
  />
</div>
`;
